Update shader compilation jank docs (#7527)
flutter/website: src/perf/shader.md

---
title: Shader compilation jank
short-title: Shader jank
description: What is shader jank and how to minimize it.
---

{% include docs/performance.md %}

If the animations on your mobile app appear to be janky, but only on the first
run, this is likely due to shader compilation. Flutter's long term solution to

> shader compilation jank is [Impeller][], which is in developer preview on master

Reviewer: ```suggestion
shader compilation jank is [Impeller][], which is in developer preview on the master
```

Author: Done.

> channel for iOS. Before continuing with the instructions below, please try out

Reviewer: ```suggestion
channel for iOS. Before continuing with the instructions below, please try
```

Author: Done.

Impeller on iOS, and let us know in a GitHub issue if it doesn't address your

> issue.

Reviewer 2: If we're not ready to encourage people to try it on Android, should we at least mention that we expect to apply a similar solution for Android?

Author: Done.

> While we work on making Impeller production ready, it is possible to mitigate

Reviewer: ```suggestion
While we work on making Impeller production ready, you can mitigate
```

shader compilation jank by bundling precompiled shaders with an iOS app.

> Unfortunately, this approach does not work well on Android due to precompiled

Reviewer: ```suggestion
Unfortunately, this approach doesn't work well on Android due to precompiled
```

Author: Done.

> shaders being device or GPU specific. The Android hardware ecosystem is diverse

Reviewer: ```suggestion
shaders being device- or GPU-specific. The Android hardware ecosystem is diverse
```

Author: Done.

> enough that the GPU specific precompiled shaders bundled with an application
> will work on only a small subset of devices, and will likely make jank worse on
> the other devices, or even create rendering errors.

Reviewer: ```suggestion
large enough that the GPU-specific precompiled shaders bundled with an application
will work on only a small subset of devices, and will likely make jank worse on
the other devices, or even create rendering errors.
```

Author: Done.

> Also please note that we are not planning to make improvements to the developer

Reviewer: ```suggestion
Also, please note that we aren't planning to make improvements to the developer
```

experience for creating precompiled shaders described below. Instead, we are
focusing our energies on the more robust solution to this problem that Impeller
offers.

## What is shader compilation jank?

A shader is a piece of code that runs on a GPU (graphics processing unit). When
the Skia graphics backend that Flutter uses for rendering sees a new sequence

> of draw commands for the first time, it will sometimes generate and compile a
> custom GPU shader for that sequence of commands. This allows that sequence and
> potentially similar sequences to render as fast as possible.

Reviewer: ```suggestion
of draw commands for the first time, it sometimes generates and compiles a
custom GPU shader for that sequence of commands. This allows that sequence and
potentially similar sequences to render as fast as possible.
```

Author: Done.

Unfortunately, Skia's shader generation and compilation happen in sequence with
the frame workload. The compilation could cost up to a few hundred milliseconds
whereas a smooth frame needs to be drawn within 16 milliseconds for a 60 fps
(frame-per-second) display. Therefore, a compilation could cause tens of frames
to be missed, and drop the fps from 60 to 6. This is _compilation jank_. After

> the compilation is complete, the animation should be smooth.

Reviewer 2: Consider adding some information here about how impeller will do this differently?

<li markdown="1"> Press `M` at the command line of `flutter run` to
    write the captured SkSL shaders into a file named something like
   `flutter_01.sksl.json`. For best results, capture SkSL shaders on an actual

>    iOS devices. A shader captured on a simulator is unlikely to work correctly
>    on actual hardware.

Reviewer: ```suggestion
   iOS device. A shader captured on a simulator isn't likely to work correctly
   on actual hardware.
```

Author: Done.

</li>

<li markdown="1"> Build the app with SkSL warm-up using the following,
    as appropriate:

```terminal
flutter build ios --bundle-sksl-path flutter_01.sksl.json
```

If it's built for a driver test like `test_driver/app.dart`, make sure to also

> specify `--target=test_driver/app.dart` (e.g., `flutter build

Reviewer: ```suggestion
specify `--target=test_driver/app.dart` (for example, `flutter build
```

Author: Done.
